content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+Postgres.app would know to find extensions in that location thanks to the
+inclusion of `$extdir_site` in the `extension_path` GUC.
+
+## Extension Directory Examples
+
+A core extension, like [citext], would live in
+`$(pg_config --extdir-core)/citext`, and have a structure such as:
+
+``` tree
+citext
+├── citext.control
+├── lib
+│   ├── citext.dylib
+│   └── bitcode
+│       ├── citext
+│       │   └── citext.bc
+│       └── citext.index.bc
+└── share
+    ├── citext--1.0--1.1.sql
+    ├── citext--1.1--1.2.sql
+    ├── citext--1.2--1.3.sql
+    ├── citext--1.3--1.4.sql
+    ├── citext--1.4--1.5.sql
+    ├── citext--1.4.sql
+    └── citext--1.5--1.6.sql
+```
+
+The subdirectory for a pure SQL extension named "pair" in a directory named
+“pair” that looks something like this:
+
+``` tree
+pair
+├── LICENSE.md
+├── README.md
+├── pair.control
+├── doc
+│   ├── html
+│   │   └── pair.html
+│   └── pair.md
+└── share
+    ├── pair--1.0--1.1.sql
+    └── pair--1.1.sql
+```
+
+A binary application like [pg_top] would live in the `pg_top` directory,
+structured something like:
+
+```
+pg_top
+├── HISTORY.rst
+├── INSTALL.rst
+├── LICENSE
+├── README.rst
+├── bin
+│   └── pg_top
+└── doc
+    └── man
+        └── man3
+            └── pg_top.3
+```
+
+And a C extension like [semver] would live in the semver directory and be
+structured something like:
+
+``` tree
+semver
+├── LICENSE
+├── README.md
+├── semver.control
+├── doc
+│   └── semver.md
+├── lib
+│   ├── semver.dylib
+│   └── bitcode
+│       ├── semver
+│       │   └── semver.bc
+│       └── semver.index.bc
+└── share
+    ├── semver--1.0--1.1.sql
+    └── semver--1.1.sql
+```
+
+## Phase Two: Preloading
+
+The above-proposed [solution](#solution) does not allow shared modules
+distributed with extensions to compatibly be loaded via [shared library
+preloading], because extension modules wil no longer live in the
+[`dynamic_library_path`]. Users can specify full paths, however. For example,
+instead of:
+
+``` ini
+shared_preload_libraries = 'pg_partman_bgw'
+```
+
+One could use:
+
+```ini
+shared_preload_libraries = '/opt/postgres/extensions/pg_partman_bgw/lib/pg_partman_bgw'
+```
+
+But users will likely find this pattern cumbersome, especially for extensions
+with multiple shared modules. Perhaps some special syntax could be added, for
+example:
+
+```ini
+shared_preload_libraries = '$extension_path:pg_partman_bgw'
+```
+
+But this overloads the semantics of `shared_preload_libraries` and friends
+rather heavily, not to mention the [`LOAD`] command.
+
+Therefore, as a follow up to the [solution](#solution) proposed above, this
+RFC proposes additional changes to PostgreSQL.
+
+### Extension Preloading
+
+Add new GUCs that complement [shared library preloading], but for *extension*
+module preloading:
+
+*   `shared_preload_extensions`
+*   `session_preload_extensions`
+*   `local_preload_extensions`
+
+Each takes a list of extensions for which to preload shared modules. In
```

Comment:
I find `local_plugins` confusing, both the name and the explanation. At least its name should be related to extensions, isn't it?

Reply:
Maybe, I was attempting to replicate the limitation on `local_preload_libraries`, which can only use modules in the "plugins" subdirectory.

Reply:
Changed it in 6a68dda; frankly I consider all the names and terms subject to change, but you're right that we don't need to keep the old term when the semantics are wrong for it.